pgxn/remotexact/funcs.c

```diff
@@ -32,22 +32,22 @@ validate_and_apply_xact(PG_FUNCTION_ARGS)
 
 	ereport(LOG, errmsg("%s", RWSetToString(rwset)));
 
-	// Mark the xact as remote before starting validation. 
-	LWLockAcquire(ProcArrayLock, LW_EXCLUSIVE);
-	MyProc->statusFlags |= PROC_IS_REMOTEXACT;
-	ProcGlobal->statusFlags[MyProc->pgxactoff] = MyProc->statusFlags;
-	LWLockRelease(ProcArrayLock);
+	/* Mark the xact as remote before starting validation by setting the
+	 * isRemoteXact flag in MyProc. We don't lock the ProcArray because its
+	 * our own process.
+	 */
+	MyProc->isRemoteXact = true;
```

Comment:
The proof for deadlock safety relies on the order of transaction upgrade and edge addition to be observed correctly by everyone. Could you add `pg_write_barrier()` after every assignment to `isRemoteXact` to make sure every memory write after that is ordered after such assignment? 
While it is very unlikely that reordering will happen at all, this rules out that possibility completely.

Reply:
Good catch; for some reason, I assumed this would be a TAS.